tests/switchless_worksleep/host/host.cpp

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <inttypes.h>
+#include <limits.h>
+#include <openenclave/host.h>
+#include <openenclave/internal/error.h>
+#include <openenclave/internal/tests.h>
+#include <openenclave/internal/thread.h>
+#include <chrono>
+#include <cstdio>
+#include <cstdlib>
+#include <thread>
+#include <vector>
+#include "../../../host/hostthread.h"
+#include "../../../host/strings.h"
+#include "switchless_worksleep_u.h"
+using namespace std;
+
+#if defined(__linux__)
+
+#include <sys/sysinfo.h>
+
+#elif defined(_WIN32)
+
+#include <Windows.h>
+#include <intrin.h>
+
+uint32_t get_nprocs(void)
+{
+    SYSTEM_INFO system_info;
+    ZeroMemory(&system_info, sizeof(system_info));
+
+    GetSystemInfo(&system_info);
+    return static_cast<size_t>(system_info.dwNumberOfProcessors);
+}
+
+#endif
+
+/**
+ * return system processors number
+ */
+int32_t get_processors_count(void)
+{
+    return get_nprocs();
+}
+
+// global counters to increment in ocalls
+uint32_t ocall1_counter = 0;
+uint32_t ocall2_counter = 0;
+
+/**
+ * ocall1 - increment counter and return
+ */
+void host_ocall1_switchless(void)
+{
+    ocall1_counter++;
+}
+
+/**
+ * ocall2 - increment counter and return
+ */
+void host_ocall2_switchless(void)
+{
+    ocall2_counter++;
+}
+
+/**
+ * thread worker method
+ * call ecall1 and ecall2 1000 times
+ * count failures
+ * sleep 1 second and repeat the above
+ */
+void* thread_func(void* arg)
+{
+    uint32_t fails = 0, i = 0;
+    oe_result_t ret = OE_OK;
+    oe_enclave_t* enclave = (oe_enclave_t*)arg;
+
+    // run ecalls
+    for (i = 0; i < 1000; i++)
+    {
+        ret = enc_ecall1_switchless(enclave);
+        if (ret != OE_OK)
+        {
+            fails++;
+        }
+        ret = enc_ecall2_switchless(enclave);
+        if (ret != OE_OK)
+        {
+            fails++;
+        }
+    }
+    OE_TEST(fails == 0);
+
+    // sleep 1 second
+    this_thread::sleep_for(std::chrono::milliseconds(1000));
```

Comment:
Why 1 second? Is the length significant? I also don't really understand why sleeping the host thread would affect the enclave thread.

Reply:
When enclave worker thread is out of work it should enter sleep mode.The sleep function is used to stop posting tasks(switchless call) and make enclave worker thread enter sleep mode.
The following code comes from  switchless implementation.
           
           // If there is no message, increment spin count until threshold is
           // reached.

            if (++context->spin_count >= spin_count_threshold)
            {
                // Reset spin count and return to host to sleep.
                context->total_spin_count += context->spin_count;
                context->spin_count = 0;
                break;
            }
            // In Release builds, the following pause has been observed to be
            // essential. Without it, the worker thread seems to hog the CPU,
            // preventing host threads from posting switchless ecall messages.
            asm volatile("pause");

Reply:
So the goal is just to release the CPU to another thread? If so I would prefer pause over sleep